In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt

In [2]:
pd.set_option(
    'display.max_colwidth', 1000
)
pd.set_option('display.max_rows', 500)
pd.options.display.max_rows = 1000
pd.options.display.max_seq_items = 2000

In [ ]:
# Look into getting more years of data, then seeing how the data roles evolved over time


In [2]:
import os

In [5]:
cwd = os.getcwd()
cwd

'F:\\stack_overflow_data\\stack_overflow_2020_data'

In [7]:
files= os.listdir(cwd)
files

['.git',
 '.ipynb_checkpoints',
 'Documentation',
 'Initial Exploration.ipynb',
 'README.md',
 'survey_results_public.csv']

In [3]:
df = pd.read_csv('survey_results_public.csv')
schema = pd.read_csv('F:\\stack_overflow_data\\stack_overflow_2020_data\\Documentation\\survey_results_schema_2020.csv')


In [10]:
df.columns

Index(['Respondent', 'MainBranch', 'Hobbyist', 'Age', 'Age1stCode', 'CompFreq',
       'CompTotal', 'ConvertedComp', 'Country', 'CurrencyDesc',
       'CurrencySymbol', 'DatabaseDesireNextYear', 'DatabaseWorkedWith',
       'DevType', 'EdLevel', 'Employment', 'Ethnicity', 'Gender', 'JobFactors',
       'JobSat', 'JobSeek', 'LanguageDesireNextYear', 'LanguageWorkedWith',
       'MiscTechDesireNextYear', 'MiscTechWorkedWith',
       'NEWCollabToolsDesireNextYear', 'NEWCollabToolsWorkedWith', 'NEWDevOps',
       'NEWDevOpsImpt', 'NEWEdImpt', 'NEWJobHunt', 'NEWJobHuntResearch',
       'NEWLearn', 'NEWOffTopic', 'NEWOnboardGood', 'NEWOtherComms',
       'NEWOvertime', 'NEWPurchaseResearch', 'NEWPurpleLink', 'NEWSOSites',
       'NEWStuck', 'OpSys', 'OrgSize', 'PlatformDesireNextYear',
       'PlatformWorkedWith', 'PurchaseWhat', 'Sexuality', 'SOAccount',
       'SOComm', 'SOPartFreq', 'SOVisitFreq', 'SurveyEase', 'SurveyLength',
       'Trans', 'UndergradMajor', 'WebframeDesireNextYear',
  

In [11]:
schema

Column  \
0                     Respondent   
1                     MainBranch   
2                       Hobbyist   
3                            Age   
4                     Age1stCode   
5                       CompFreq   
6                      CompTotal   
7                  ConvertedComp   
8                        Country   
9                   CurrencyDesc   
10                CurrencySymbol   
11        DatabaseDesireNextYear   
12            DatabaseWorkedWith   
13                       DevType   
14                       EdLevel   
15                    Employment   
16                     Ethnicity   
17                        Gender   
18                    JobFactors   
19                        JobSat   
20                       JobSeek   
21        LanguageDesireNextYear   
22            LanguageWorkedWith   
23        MiscTechDesireNextYear   
24            MiscTechWorkedWith   
25  NEWCollabToolsDesireNextYear   
26      NEWCollabToolsWorkedWith   
27                     NEWDevOps   
28                 NEWDevOpsImpt   
29                     NEWEdImpt   
30                    NEWJobHunt   
31            NEWJobHuntResearch   
32                      NEWLearn   
33                   NEWOffTopic   
34                NEWOnboardGood   
35                 NEWOtherComms   
36                   NEWOvertime   
37           NEWPurchaseResearch   
38                 NEWPurpleLink   
39                    NEWSOSites   
40                      NEWStuck   
41                         OpSys   
42                       OrgSize   
43        PlatformDesireNextYear   
44            PlatformWorkedWith   
45                  PurchaseWhat   
46                     Sexuality   
47                     SOAccount   
48                        SOComm   
49                    SOPartFreq   
50                   SOVisitFreq   
51                    SurveyEase   
52                  SurveyLength   
53                         Trans   
54                UndergradMajor   
55        WebframeDesireNextYear   
56            WebframeWorkedWith   
57                 WelcomeChange   
58                   WorkWeekHrs   
59                     YearsCode   
60                  YearsCodePro   

                                                                                                                                                                                                                                                                                                                                        QuestionText  
0                                                                                                                                                                                                                                                                             Randomized respondent ID number (not in order of survey response time)  
1                                                                                                                                                                                                                                   Which of the following options best describes you today? Here, by "developer" we mean "someone who writes code."  
2                                                                                                                                                                                                                                                                                                                            Do you code as a hobby?  
3                                                                                                                                                                                                                                                       What is your age (in years)? If you prefer not to answer, you may leave this question blank.  
4                                                                             

In [4]:
schema.loc[schema['Column'] == 'CompTotal']

Column                                       QuestionText
6  CompTotal  What is your current total compensation (salar...

In [5]:
schema.loc[schema['Column'] == 'ConvertedComp']

Column                                       QuestionText
7  ConvertedComp  Salary converted to annual USD salaries using ...

In [6]:
df['DatabaseDesireNextYear'].value_counts()

PostgreSQL                                                         2310
Microsoft SQL Server                                               1947
MySQL                                                              1823
MongoDB                                                            1624
SQLite                                                              966
                                                                   ... 
Cassandra;Elasticsearch;MariaDB;MongoDB;MySQL;Oracle;PostgreSQL       1
Couchbase;DynamoDB;Microsoft SQL Server;Oracle;SQLite                 1
Cassandra;Elasticsearch;IBM DB2;PostgreSQL;Redis                      1
Elasticsearch;IBM DB2;MariaDB;MongoDB;Oracle                          1
Cassandra;Elasticsearch;Firebase;MariaDB;MongoDB;Redis;SQLite         1
Name: DatabaseDesireNextYear, Length: 3193, dtype: int64

In [7]:
dev_types = df['DevType'].value_counts().index.to_list()

In [8]:
# For now, these are the roles I'm interested in
# Do I want to add database administrator? Maybe later.
data_roles = [
    'Data scientist or machine learning specialist',
    'Data or business analyst',
    'Engineer, data',
]

In [16]:
dev_types

['Developer, full-stack',
 'Developer, back-end',
 'Developer, back-end;Developer, front-end;Developer, full-stack',
 'Developer, back-end;Developer, full-stack',
 'Developer, front-end',
 'Developer, mobile',
 'Developer, front-end;Developer, full-stack',
 'Developer, back-end;Developer, desktop or enterprise applications;Developer, front-end;Developer, full-stack',
 'Developer, back-end;Developer, desktop or enterprise applications',
 'Developer, desktop or enterprise applications',
 'Developer, back-end;Developer, front-end;Developer, full-stack;Developer, mobile',
 'Developer, back-end;Developer, desktop or enterprise applications;Developer, full-stack',
 'Developer, embedded applications or devices',
 'Data scientist or machine learning specialist',
 'Developer, full-stack;Developer, mobile',
 'Developer, back-end;Developer, front-end',
 'Developer, desktop or enterprise applications;Developer, full-stack',
 'Data or business analyst',
 'Developer, QA or test',
 'Developer, front-

In [9]:
# I need to write a loop to grab things in the dataframe
# for each role in the list, then concatenate them all, I think.
ds_df = df.loc[df['DevType'].str.contains('|'.join(data_roles), na=False)]
print(len(df))
print(len(ds_df))

64461
8726


In [10]:
ds_df['DevType'].value_counts()

Data scientist or machine learning specialist                                                                                                                                                                                                                                                             369
Data or business analyst                                                                                                                                                                                                                                                                                  285
Developer, back-end;Engineer, data                                                                                                                                                                                                                                                                        227
Engineer, data                                                                                

## Columns to Investigate

- Age
- ConvertedComp
- Country
- DatabaseWorkedWith
- EdLevel
- Employment (as a filter)
- JobSat
- LanguageWorkedWith
- LanguageDesireNextYear
- MiscTechWorkedWith
- NEWLearn
- NEWOvertime
- NEWStuck
- OpSys
- OrgSize
- UndergradMajor
- WorkWeekHrs
- YearsCode
- YearsCodePro (may be able to drop this later)

## Questions I want to answer:

- What career track is the most satisfied overall: Data Scientist, Data Engineers, or Data analysts?
- Where in the world is compensation the best for data positions?
- Which languages and databases are used by the highest-paid data scientists/engineers/analysts?
- Which countries seem to have the best work-life balance (WorkWeekHrs/NEWOvertime)?
- Do respondents who use R get paid more or less than respondents that use Python, or are they roughly equivalent?

In [11]:
df = ds_df[[
    'DevType',
    'Age',
    'CompTotal',
    'ConvertedComp',
#     'CompFreq', # Don't really care about this.
    'Country',
    'DatabaseWorkedWith',
    'EdLevel',
    'Employment', 
    'JobSat',
    'LanguageWorkedWith',
#     'LanguageDesireNextYear',
#     'MiscTechWorkedWith', # Lots of NaNs here. We can come back to this.
#     'NEWLearn',
    'NEWOvertime',
    'NEWStuck', # Could be useful to see which strategies are used by whom
    'OpSys',
    'OrgSize',
    'UndergradMajor',
#     'WorkWeekHrs',
    'YearsCode',
#     'YearsCodePro'
]]

In [12]:
df.head(1)

DevType  Age  CompTotal  ConvertedComp  \
21  Developer, full-stack;Engineer, data  NaN        NaN            NaN   

   Country DatabaseWorkedWith                                       EdLevel  \
21   India                NaN  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

            Employment                 JobSat LanguageWorkedWith  \
21  Employed full-time  Slightly dissatisfied        Java;Python   

                         NEWOvertime NEWStuck    OpSys               OrgSize  \
21  Often: 1-2 days per week or more      NaN  Windows  500 to 999 employees   

                                       UndergradMajor YearsCode  
21  Computer science, computer engineering, or sof...        10

In [13]:
df.Employment.value_counts()

Employed full-time                                      6904
Independent contractor, freelancer, or self-employed    1080
Not employed, but looking for work                       388
Employed part-time                                       354
Name: Employment, dtype: int64

I want to get rid of the 388 responses from people who are unemployed. I'm not sure what I want to do with part-timers, though... For now, let's drop them.

In [14]:
employed_df = df[
    (df['Employment'] != 'Not employed, but looking for work') &
    (df['Employment'] != 'Employed part-time')
]
employed_df.Employment.value_counts()

Employed full-time                                      6904
Independent contractor, freelancer, or self-employed    1080
Name: Employment, dtype: int64

We successfully dropped the respondents that weren't employed or were workign part-time.

In [15]:
employed_df.dtypes

DevType                object
Age                   float64
CompTotal             float64
ConvertedComp         float64
Country                object
DatabaseWorkedWith     object
EdLevel                object
Employment             object
JobSat                 object
LanguageWorkedWith     object
NEWOvertime            object
NEWStuck               object
OpSys                  object
OrgSize                object
UndergradMajor         object
YearsCode              object
dtype: object

Our only numeric columns are age, ConvertedComp, CompTotal, and WorkWeekHrs. If the count of NaNs is relatively similar between ConvertedComp and CompTotal, we can probably drop CompTotal, too.

In [16]:
for col in employed_df.columns.tolist():
    prop_sals = 1-employed_df.isnull()[col].mean()
    print(f'Proportion of values reported for {col}: {prop_sals}')

Proportion of values reported for DevType: 1.0
Proportion of values reported for Age: 0.7813126252505009
Proportion of values reported for CompTotal: 0.7050350701402806
Proportion of values reported for ConvertedComp: 0.7036573146292585
Proportion of values reported for Country: 1.0
Proportion of values reported for DatabaseWorkedWith: 0.846317635270541
Proportion of values reported for EdLevel: 0.9847194388777555
Proportion of values reported for Employment: 1.0
Proportion of values reported for JobSat: 0.9347444889779559
Proportion of values reported for LanguageWorkedWith: 0.9407565130260521
Proportion of values reported for NEWOvertime: 0.8976703406813628
Proportion of values reported for NEWStuck: 0.9222194388777555
Proportion of values reported for OpSys: 0.9319889779559118
Proportion of values reported for OrgSize: 0.9182114228456913
Proportion of values reported for UndergradMajor: 0.9304859719438878
Proportion of values reported for YearsCode: 0.9901052104208417


The number of NaNs appear to be roughly the same between CompTotal and ConvertedComp (~0.70), so we can drop CompTotal.

Overall, the response rate within this group, and for the fields I selected, is pretty great! The lowest response rates are salary-related. As a result, I may want to use different copies of the data set for different questions. There's no point in throwing away extra data points for answering questions not related to salary.

In [17]:
employed_df.drop(columns=['CompTotal'], inplace=True)

c:\users\stew\miniconda3\lib\site-packages\pandas\core\frame.py:4167: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


I can't interpolate missing values for non-numeric columns, so I can just drop those.

It doesn't make sense to interpolate age or work week hours. For the latter, I could do the mode, but I don't know that it would shake out the same for all countries represented.

In [18]:
print(len(employed_df))
filt_df = employed_df.dropna(how='any')
print(len(filt_df))

7984
3975


Ok, that's roughly half of the original data set. Maybe I should go back and narrow down my columns before taking this step.

In [19]:
# Check max salary. $2 million annually? This seems unlikely.
print((employed_df['ConvertedComp'].max()))

2000000.0


In [20]:
employed_df[employed_df['ConvertedComp'] == 2000000.0]

DevType   Age  ConvertedComp  \
123    Data scientist or machine learning specialist;...  41.0      2000000.0   
722    Developer, back-end;Engineer, data;System admi...   NaN      2000000.0   
816                             Data or business analyst  40.0      2000000.0   
1581   Data or business analyst;Data scientist or mac...  59.0      2000000.0   
1616   Data or business analyst;Database administrato...  48.0      2000000.0   
2556      Data or business analyst;Developer, full-stack  40.0      2000000.0   
2670   Data or business analyst;Data scientist or mac...  23.0      2000000.0   
2761   Data or business analyst;Data scientist or mac...  45.0      2000000.0   
3058   Data scientist or machine learning specialist;...  32.0      2000000.0   
3170   Data scientist or machine learning specialist;...  25.0      2000000.0   
3247   Data scientist or machine learning specialist;...  34.0      2000000.0   
3265   Data or business analyst;Database administrato...  43.0      2000000.0   
3932   Data or business analyst;Data scientist or mac...  25.0      2000000.0   
4011   Academic researcher;Data scientist or machine ...  29.0      2000000.0   
4109      Data or business analyst;Developer, full-stack  24.0      2000000.0   
4113   Data or business analyst;Data scientist or mac...  23.0      2000000.0   
9802   Academic researcher;Data or business analyst;D...  47.0      2000000.0   
14062  Database administrator;Developer, back-end;Dev...  41.0      2000000.0   
15229  Data or business analyst;Database administrato...  49.0      2000000.0   
16769  Data or business analyst;Data scientist or mac...  39.0      2000000.0   
17288  Data or business analyst;Developer, desktop or...  23.0      2000000.0   
18126      Data scientist or machine learning specialist  28.0      2000000.0   
20563  Data scientist or machine learning specialist;...  26.0      2000000.0   
24294  Data or business analyst;Data scientist or mac...  32.0      2000000.0   
24756  Data or business analyst;Developer, back-end;D...  33.0      2000000.0   
25045  Data or business analyst;Data scientist or mac...  25.0      2000000.0   
25088  Data or business analyst;Data scientist or mac...  30.0      2000000.0   
25552  Data or business analyst;Database administrato...  39.0      2000000.0   
29250  Data scientist or machine learning specialist;...  46.0      2000000.0   
29294  Data or business analyst;Data scientist or mac...  39.0      2000000.0   
29824  Data or business analyst;Data scientist or mac...  27.0      2000000.0   
33134  Data or business analyst;Database administrato...  38.0      2000000.0   
33664  Data or business analyst;Data scientist or mac...  28.0      2000000.0   
35569  Database administrator;Developer, embedded app...   NaN      2000000.0   
39766  Data or business analyst;Data scientist or mac...   NaN      2000000.0   
39912                 Developer, back-end;Engineer, data  26.0      2000000.0   
45753  Academic researcher;Data or business analyst;D...   NaN      2000000.0   
46965  Data or business analyst;Data scientist or mac...  34.0      2000000.0   
47111  Academic researcher;Data or business analyst;D...   NaN      2000000.0   
47357  Data or business analyst;Designer;Developer, b...  26.0      2000000.0   
47462  Data or business analyst;Database administrato...  46.0      2000000.0   
47550  Data scientist or machine learning specialist;...  47.0      2000000.0   
48436  Data or business analyst;Data scientist or mac...   NaN      2000000.0   
48692  Data or business analyst;Data scientist or mac...  26.0      2000000.0   
49178  Engineer, data;Engineer, site reliability;Prod...  27.0      2000000.0   
49854  Data scientist or machine learning specialist;...   NaN      2000000.0   
50951  Data scientist or machine learning specialist;...  28.0      2000000.0   
50965                     Engineer, data;Product manager  26.0      2000000.0   
51123  Developer, back-end;Developer, desktop or ente...  30.0      2000000.0   

In [21]:
print(len(employed_df.Country.value_counts()))

144


There are way too many countries for me to use this as a categorical variable. I could:

- Group them by continent (I could use a library like pycountry-convert)
- Eliminate any countries with fewer than n values
- Only keep countries I've thought of visiting before

I need to write a function that will:

- Separate the job title strings into elements of a list
- Create binary columns for Data Analyst, Data Scientist, and Data engineer
- Loop through the df and add a 1 in those columns if present, 0 if not.

In [29]:
dev_list = employed_df['DevType'].tolist()
dev_list[0]

'Developer, full-stack;Engineer, data'

In [32]:
employed_df.dropna(subset=['DevType'], axis = 0, inplace=True)

<ipython-input-32-dabb0a2014f9>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  employed_df.dropna(subset=['DevType'], axis = 0, inplace=True)


In [33]:
ds_string = 'Data scientist or machine learning specialist'
da_string = 'Data or business analyst'
de_string = 'Engineer, data'

In [38]:
employed_df.loc[
    employed_df["DevType"].str.contains(ds_string) == True,
    'DataScientist'
] = 1
employed_df['DataScientist'].fillna(value=0, axis=0, inplace=True)

employed_df.loc[
    employed_df["DevType"].str.contains(da_string) == True,
    'DataAnalyst'
] = 1
employed_df['DataAnalyst'].fillna(value=0, axis=0, inplace=True)

employed_df.loc[
    employed_df["DevType"].str.contains(de_string) == True,
    'DataEngineer'
] = 1
employed_df['DataEngineer'].fillna(value=0, axis=0, inplace=True)

employed_df.head()

c:\users\stew\miniconda3\lib\site-packages\pandas\core\indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)
c:\users\stew\miniconda3\lib\site-packages\pandas\core\series.py:4530: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().fillna(
c:\users\stew\miniconda3\lib\site-packages\pandas\core\indexing.py:1763: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pand

DevType   Age  ConvertedComp  \
21               Developer, full-stack;Engineer, data   NaN            NaN   
24  Developer, back-end;Developer, full-stack;DevO...   NaN            NaN   
29  Data or business analyst;Database administrato...   NaN        38778.0   
35  Data or business analyst;Data scientist or mac...  34.0        77556.0   
43  Data or business analyst;Developer, back-end;P...  32.0        55893.0   

           Country                           DatabaseWorkedWith  \
21           India                                          NaN   
24        Portugal                                       Oracle   
29  United Kingdom            Microsoft SQL Server;Redis;SQLite   
35  United Kingdom  IBM DB2;Microsoft SQL Server;MongoDB;SQLite   
43          Brazil                         Microsoft SQL Server   

                                              EdLevel          Employment  \
21       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  Employed full-time   
24                Associate degree (A.A., A.S., etc.)  Employed full-time   
29       Bachelor’s degree (B.A., B.S., B.Eng., etc.)  Employed full-time   
35  Some college/university study without earning ...  Employed full-time   
43    Master’s degree (M.A., M.S., M.Eng., MBA, etc.)  Employed full-time   

                                JobSat  \
21               Slightly dissatisfied   
24  Neither satisfied nor dissatisfied   
29               Slightly dissatisfied   
35                  Slightly satisfied   
43  Neither satisfied nor dissatisfied   

                                   LanguageWorkedWith  \
21                                        Java;Python   
24                            Bash/Shell/PowerShell;C   
29  Bash/Shell/PowerShell;C#;HTML/CSS;JavaScript;P...   
35        C#;Go;HTML/CSS;Java;JavaScript;Python;R;SQL   
43                          HTML/CSS;Python;R;SQL;VBA   

                                          NEWOvertime  \
21                   Often: 1-2 days per week or more   
24                   Often: 1-2 days per week or more   
29  Occasionally: 1-2 days per quarter but less th...   
35  Sometimes: 1-2 days per month but less than we...   
43                   Often: 1-2 days per week or more   

                                             NEWStuck    OpSys  \
21                                                NaN  Windows   
24                               Visit Stack Overflow  Windows   
29  Play games;Call a coworker or friend;Visit Sta...  Windows   
35  Visit Stack Overflow;Go for a walk or other ph...  Windows   
43  Call a coworker or friend;Visit Stack Overflow...  Windows   

                     OrgSize  \
21      500 to 999 employees   
24      100 to 499 employees   
29                       NaN   
35  1,000 to 4,999 employees   
43        10 to 19 employees   

                                       UndergradMajor YearsCode  \
21  Computer science, computer engineering, or sof...        10   
24  Computer science, computer engineering, or sof...        23   
29  Information systems, information technology, o...         4   
35  Computer science, computer engineering, or sof...         4   
43  Another engineering discipline (such as civil,...        10   

    DataScientist  DataAnalyst  DataEngineer  
21            0.0          0.0           1.0  
24            0.0          0.0           1.0  
29            0.0          1.0           0.0  
35            1.0          1.0           0.0  
43            0.0          1.0           0.0

In [39]:
# Now we can drop 'DevType'

employed_df.drop(columns=['DevType'], inplace=True)

c:\users\stew\miniconda3\lib\site-packages\pandas\core\frame.py:4167: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


In [27]:
# # Debug this
# def job_type_simplifier(string):
#     '''
#     INPUT
#     df - pandas dataframe with 2020 SO survey data
    
#     OUTPUT
#     job_df - pandas dataframe with three new columns:
#         - DataAnalyst (values 0 or 1)
#         - DataScientist (values 0 or 1)
#         - DataEngineer (values 0 or 1)
#     and drops the DevType column.
#     The steps are: 1. Drop NaNs from DevType. 2. Initialize new columns.
#     3. Loop through dataframe, separate the DevType string on the ';'
#     character, and add each element to a list. Check that list for each of
#     the job types of interest, and fill in the appropriate column accordingly.
#     '''
#     type_list = string.split(';').tolist()
#     print(len(type_list))
    
#     if 'Data scientist or machine learning specialist' in type_list == True:
#         DataScientist.append(1)
#     else:
#         DataScientist.append(0)
#     if 'Data or business analyst' in type_list == True:
#          DataAnalyst.append(1)
#     else:
#         DataAnalyst.append(0)
#     if 'Engineer, data' in type_list == True:
#          DataEngineer.append(1)
#     else:
#         DataEngineer.append(0)
 
    
#     return new_df